// @nx/angular@19.8.4/src/generators/library/lib/add-project.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addProject = addProject;
const devkit_1 = require("@nx/devkit");
const target_defaults_utils_1 = require("@nx/devkit/src/generators/target-defaults-utils");
function addProject(tree, libraryOptions) {
    const project = {
        name: libraryOptions.name,
        root: libraryOptions.projectRoot,
        sourceRoot: (0, devkit_1.joinPathFragments)(libraryOptions.projectRoot, 'src'),
        prefix: libraryOptions.prefix,
        tags: libraryOptions.parsedTags,
        projectType: 'library',
        targets: {},
    };
    if (libraryOptions.buildable || libraryOptions.publishable) {
        const executor = libraryOptions.publishable
            ? '@nx/angular:package'
            : '@nx/angular:ng-packagr-lite';
        (0, target_defaults_utils_1.addBuildTargetDefaults)(tree, executor);
        project.targets.build = {
            executor,
            outputs: ['{workspaceRoot}/dist/{projectRoot}'],
            options: {
                project: `${libraryOptions.projectRoot}/ng-package.json`,
            },
            configurations: {
                production: {
                    tsConfig: `${libraryOptions.projectRoot}/tsconfig.lib.prod.json`,
                },
                development: {
                    tsConfig: `${libraryOptions.projectRoot}/tsconfig.lib.json`,
                },
            },
            defaultConfiguration: 'production',
        };
    }
    (0, devkit_1.addProjectConfiguration)(tree, libraryOptions.name, project);
    return project;
}
