// @nx/vite@20.1.4/src/utils/e2e-web-server-info-utils.js
"use strict";
Object.defineProperty(exports, "getViteE2EWebServerInfo", {
    enumerable: true,
    get: function() {
        return getViteE2EWebServerInfo;
    }
});
const _devkit = require("@nx/devkit");
const _e2ewebserverinfoutils = require("@nx/devkit/src/generators/e2e-web-server-info-utils");
async function getViteE2EWebServerInfo(tree, projectName, configFilePath, isPluginBeingAdded, e2ePortOverride) {
    var _nxJson_targetDefaults, _nxJson_targetDefaults_serve_options, _nxJson_targetDefaults1;
    const nxJson = (0, _devkit.readNxJson)(tree);
    let e2ePort = e2ePortOverride != null ? e2ePortOverride : 4200;
    if (((_nxJson_targetDefaults = nxJson.targetDefaults) == null ? void 0 : _nxJson_targetDefaults['serve']) && ((_nxJson_targetDefaults1 = nxJson.targetDefaults) == null ? void 0 : (_nxJson_targetDefaults_serve_options = _nxJson_targetDefaults1['serve'].options) == null ? void 0 : _nxJson_targetDefaults_serve_options.port)) {
        var _nxJson_targetDefaults_serve_options1, _nxJson_targetDefaults2;
        e2ePort = (_nxJson_targetDefaults2 = nxJson.targetDefaults) == null ? void 0 : (_nxJson_targetDefaults_serve_options1 = _nxJson_targetDefaults2['serve'].options) == null ? void 0 : _nxJson_targetDefaults_serve_options1.port;
    }
    return (0, _e2ewebserverinfoutils.getE2EWebServerInfo)(tree, projectName, {
        plugin: '@nx/vite/plugin',
        serveTargetName: 'serveTargetName',
        serveStaticTargetName: 'previewTargetName',
        configFilePath
    }, {
        defaultServeTargetName: 'serve',
        defaultServeStaticTargetName: 'preview',
        defaultE2EWebServerAddress: `http://localhost:${e2ePort}`,
        defaultE2ECiBaseUrl: 'http://localhost:4300',
        defaultE2EPort: e2ePort
    }, isPluginBeingAdded);
}
