// @nx/angular@19.8.4/src/generators/setup-mf/lib/update-host-app-routes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateHostAppRoutes = updateHostAppRoutes;
const devkit_1 = require("@nx/devkit");
const route_utils_1 = require("../../../utils/nx-devkit/route-utils");
const version_utils_1 = require("../../utils/version-utils");
function updateHostAppRoutes(tree, options) {
    const { sourceRoot } = (0, devkit_1.readProjectConfiguration)(tree, options.appName);
    tree.write((0, devkit_1.joinPathFragments)(sourceRoot, 'app/app.component.html'), `<ul class="remote-menu">
<li><a routerLink="/">Home</a></li>
</ul>
<router-outlet></router-outlet>
`);
    let pathToHostRootRoutingFile = (0, devkit_1.joinPathFragments)(sourceRoot, 'app/app.routes.ts');
    let hostRootRoutingFile = tree.read(pathToHostRootRoutingFile, 'utf-8');
    if (!hostRootRoutingFile) {
        pathToHostRootRoutingFile = (0, devkit_1.joinPathFragments)(sourceRoot, 'app/app-routing.module.ts');
        hostRootRoutingFile = tree.read(pathToHostRootRoutingFile, 'utf-8');
    }
    (0, route_utils_1.addRoute)(tree, pathToHostRootRoutingFile, `{
      path: '',
      component: NxWelcomeComponent
    }`);
    tree.write(pathToHostRootRoutingFile, `import { NxWelcomeComponent } from './nx-welcome.component';
${tree.read(pathToHostRootRoutingFile, 'utf-8')}`);
    const { major: angularMajorVersion } = (0, version_utils_1.getInstalledAngularVersionInfo)(tree);
    (0, devkit_1.generateFiles)(tree, (0, devkit_1.joinPathFragments)(__dirname, '../files/host-files'), (0, devkit_1.joinPathFragments)(sourceRoot, 'app'), {
        appName: options.appName,
        standalone: options.standalone,
        useRouterTestingModule: angularMajorVersion < 18,
        tmpl: '',
    });
}
